### PR TITLE
GH-34 db sync time

### PR DESCRIPTION
**Pull request by danielmkn**
- Adding one more resource `resource_xray_settings`. It will set Xray DB Sync time for now, with possible expansion for other settings. Thus the name is generic, `settings`, not `db_sync`. 
-DB Sync Time API doesn't have create and delete functionality. See the [docs](https://www.jfrog.com/confluence/display/JFROG/Xray+REST+API#XrayRESTAPI-UpdateDBSyncDailyUpdateTime).
- Since we must implement all CRUD operations in the provider, I used a workaround - delete function just removes the object from the TF state, create function does the update call. 
- Sync time is set by default during the Xray installation, so there is no need to create it and no need to delete it. 
- Documentation was auto-generated.

### CODE REVIEW

File: docs/resources/settings.md [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+<!-- schema generated by tfplugindocs -->
+## Schema
+
+### Required
+
+- **db_sync_updates_time** (String) The time of the Xray DB sync daily update job.
+
+### Optional
+
+- **id** (String) The ID of this resource.
```
Comment (chb0github) [inline docs/resources/settings.md:30]: mmm, no. Not optional - also not managable by them

File: pkg/xray/resource_xray_settings.go [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+
+// No create functionality provided by API. DB sync time exists by default and can only be modified
+func resourceXrayDbSyncTimeCreate(ctx context.Context, d *schema.ResourceData, m interface{}) diag.Diagnostics {
+	return resourceXrayDbSyncTimeUpdate(ctx, d, m)
+}
+
+func resourceXrayDbSyncTimeRead(ctx context.Context, d *schema.ResourceData, m interface{}) diag.Diagnostics {
+	dbSyncTime, resp, err := getDbSyncTime(m.(*resty.Client))
+	if err != nil {
+		if resp != nil && resp.StatusCode() == http.StatusNotFound {
```
Comment (chb0github) [inline pkg/xray/resource_xray_settings.go:69]: This is not possible. If this occurs it's permanently broken. The client may have misconfigured something on their end. it's definitely an error and in your description, make it clear how wild of an error this must be

File: docs/resources/settings.md (hunk 1 above)
Comment (danielmkn) [reply docs/resources/settings.md:30]: for some reason, that field was generated by `tfgeneratedocs` plugin. I'll remove it

File: pkg/xray/resource_xray_settings.go [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+}
+
+func packDBSyncTime(dbSyncTime DbSyncDailyUpdatesTime, d *schema.ResourceData) diag.Diagnostics {
+	if err := d.Set("db_sync_updates_time", dbSyncTime.DbSyncTime); err != nil {
+		return diag.FromErr(err)
+	}
+	return nil
+}
+
+func getDbSyncTime(client *resty.Client) (DbSyncDailyUpdatesTime, *resty.Response, error) {
```
Comment (chb0github) [inline pkg/xray/resource_xray_settings.go:55]: You created this function and then use it in 1 place? You've actually added 3 lines of code. Please remove this function and inline it

File: pkg/xray/resource_xray_settings.go [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+	dbSyncTime := DbSyncDailyUpdatesTime{}
+	resp, err := client.R().SetResult(&dbSyncTime).Get("xray/api/v1/configuration/dbsync/time")
+	return dbSyncTime, resp, err
+}
+
+// No create functionality provided by API. DB sync time exists by default and can only be modified
+func resourceXrayDbSyncTimeCreate(ctx context.Context, d *schema.ResourceData, m interface{}) diag.Diagnostics {
+	return resourceXrayDbSyncTimeUpdate(ctx, d, m)
```
Comment (chb0github) [inline pkg/xray/resource_xray_settings.go:63]: Don't do this - literally in the schema config use the same function for both. You don't need this function at all

File: CHANGELOG.md [hunk 5]
```diff
@@ -1,3 +1,10 @@
+## 0.0.4 (Mar 16, 2022)
```
[the remark this answers is not in the document]
Comment (danielmkn) [reply CHANGELOG.md:1]: Change to 1.0.0?

File: pkg/xray/resource_xray_settings.go [hunk 6]
```diff
@@ -0,0 +1,99 @@
...
+				Type:             schema.TypeString,
+				Required:         true,
+				Description:      "The time of the Xray DB sync daily update job.",
+				ValidateDiagFunc: matchesHoursMinutesTime,
+			},
+		},
+	}
+}
+
+type DbSyncDailyUpdatesTime struct {
```
[the remark this answers is not in the document]
Comment (danielmkn) [reply pkg/xray/resource_xray_settings.go:36]: I love the struct

File: pkg/xray/validators.go [hunk 7]
```diff
@@ -533,3 +533,7 @@ func minLength(length int) func(i interface{}, k string) ([]string, []error) {
 func inList(strings ...string) schema.SchemaValidateDiagFunc {
 	return validation.ToDiagFunc(validation.StringInSlice(strings, true))
 }
+
+var matchesHoursMinutesTime = validation.ToDiagFunc(
+	validation.StringMatch(regexp.MustCompile(`^([0-1]?[0-9]|2[0-3]):[0-5][0-9]$`), "Wrong format input, expected valid hour:minutes form"),
```
Comment (chb0github) [inline pkg/xray/validators.go:538]: Wait a minute - this can't be right

SO, `9:00` is ok? Is that AM or PM? 

You're not testing your time validation sufficiently. Please test:

- [ ] `09:00` -`09:59` pass
- [ ] `h:mm` fails
- [ ] `00:00` - `09:59` pass
- [ ] Anything greater than 23:59 fails


The rest of the regex looks ok, but you need some more edge case testing. For a format of `HH:mm` it has to be in 24h format. If needs be, you can break this down into a few regex for clarity and use `validation.all` to combine them. Also, your error message isn't specific enough. It needs: `HH:mm`

File: pkg/xray/resource_xray_settings.go (hunk 4 above)
Comment (danielmkn) [reply pkg/xray/resource_xray_settings.go:63]: ok, removed. My thought was it will be more clear what and why we are doing it

File: pkg/xray/resource_xray_settings.go [hunk 8]
```diff
@@ -0,0 +1,99 @@
...
+
+		Importer: &schema.ResourceImporter{
+			StateContext: schema.ImportStatePassthroughContext,
+		},
+
+		Schema: map[string]*schema.Schema{
+			"db_sync_updates_time": {
+				Type:             schema.TypeString,
+				Required:         true,
+				Description:      "The time of the Xray DB sync daily update job.",
```
Comment (chb0github) [inline pkg/xray/resource_xray_settings.go:29]: In the description, give the exact format

**Review by chb0github — CHANGES_REQUESTED** [review]
Some changes are required

File: pkg/xray/resource_xray_settings.go (hunk 8 above)
Comment (danielmkn) [reply pkg/xray/resource_xray_settings.go:29]: Fixed

File: pkg/xray/resource_xray_settings.go (hunk 2 above)
Comment (danielmkn) [reply pkg/xray/resource_xray_settings.go:69]: Changed

File: pkg/xray/resource_xray_settings.go (hunk 3 above)
Comment (danielmkn) [reply pkg/xray/resource_xray_settings.go:55]: removed

File: pkg/xray/validators.go (hunk 7 above)
Comment (danielmkn) [reply pkg/xray/validators.go:538]: `9:00` is ok by Xray. If set like this, it's returned as-is. It's 24 hrs format, but for some reason missing 0 is ok. The format description was added to the docs. 
The error message is what API returns, but I'll add the `HH:mm` format mention.
Added several test scenarios.

File: pkg/xray/resource_xray_settings.go [hunk 9]
```diff
@@ -0,0 +1,88 @@
...
+	packDBSyncTime(dbSyncTime, d)
+	return nil
+}
+
+func resourceXrayDbSyncTimeUpdate(ctx context.Context, d *schema.ResourceData, m interface{}) diag.Diagnostics {
+	dbSyncTime := unpackDBSyncTime(d)
+	resp, err := m.(*resty.Client).R().SetBody(dbSyncTime).Put("xray/api/v1/configuration/dbsync/time")
+	if err != nil {
+		if resp != nil && resp.StatusCode() == http.StatusNotFound {
+			log.Printf("[WARN] DB sync settings (%s) not found, removing from state", d.Id())
```
Comment (chb0github) [inline pkg/xray/resource_xray_settings.go:73]: This is still an error - if you can't actually delete this resource then it should be absolutely impossible to get 404. This is an error

File: pkg/xray/resource_xray_settings_test.go [hunk 10]
```diff
@@ -0,0 +1,50 @@
...
+				Config: dbSyncTime(resourceName, time),
+				Check:  resource.TestCheckResourceAttr(fqrn, "db_sync_updates_time", time),
+			},
+		},
+	})
+}
+
+func TestDbSyncTimeNegative(t *testing.T) {
+	_, _, resourceName := mkNames("db_sync-", "xray_settings")
+	var invalidTime = []string{"24:00", "24:55", "", "12:0", "string", "12pm"}
```
Comment (chb0github) [inline pkg/xray/resource_xray_settings_test.go:29]: `9:00` must also be rejected. I know xray will accept it, but it's not proper 24h time

**Review by chb0github — CHANGES_REQUESTED** [review]

File: pkg/xray/resource_xray_settings_test.go (hunk 10 above)
Comment (danielmkn) [reply pkg/xray/resource_xray_settings_test.go:29]: I think we need to accept everything Xray will accept.

Comment (danielmkn) [reply pkg/xray/resource_xray_settings_test.go:29]: I've changed regex to reject H:mm. Updated the test

File: pkg/xray/resource_xray_settings.go (hunk 9 above)
Comment (danielmkn) [reply pkg/xray/resource_xray_settings.go:73]: Updated to reject H:mm

Comment (danielmkn) [reply pkg/xray/resource_xray_settings.go:73]: Do I remove it from the state on update error? Regex verification won't allow API call to error out on the wrong time format. So the only possibility, as you said, it a very broken installation.

Comment (danielmkn) [reply pkg/xray/resource_xray_settings.go:73]: fixed

File: pkg/xray/validators.go (hunk 7 above)
Comment (danielmkn) [reply pkg/xray/validators.go]: I don't get it - it is enforced

**Review by chb0github — APPROVED** [review]
